Tasks/Grunt/grunttask.ts

```diff
@@ -41,11 +40,83 @@ gt.pathArg(gruntFile);
 
 gt.argString(tl.getInput('arguments', false));
 
+var isCodeCoverageEnabled = tl.getBoolInput('enableCodeCoverage');
+
+if (isCodeCoverageEnabled) {
+	var npm = tl.createToolRunner(tl.which('npm', true));
+	npm.argString('install istanbul');
+	
+	var testSrc = tl.getPathInput('testFiles', true, false);
+
+	if (os.type().match(/^Win/)) {
+		var istanbul = tl.createToolRunner(tl.which('istanbul', true));
+	} else {
+		var istanbul = tl.createToolRunner(tl.which('./node_modules/istanbul/lib/cli.js', true));
+	}
+	istanbul.argString('cover --report cobertura --report html');
+	istanbul.arg('./node_modules/mocha/bin/_mocha');
+	istanbul.arg(testSrc);
+	istanbul.argString('-- --ui bdd -t 5000');
+
+	var buildFolder = tl.getVariable('System.DefaultWorkingDirectory');
+	var summaryFile = path.join(buildFolder, 'coverage/cobertura-coverage.xml');
+	var reportDirectory = path.join(buildFolder, 'coverage/');
+}
+
+var publishJUnitResults = tl.getBoolInput('publishJUnitResults');
+var testResultsFiles = tl.getInput('testResultsFiles', publishJUnitResults);
+
 gt.exec()
-.then(function(code) {
-	tl.setResult(tl.TaskResult.Succeeded, tl.loc('GruntReturnCode', code));
-})
-.fail(function(err) {
-	tl.debug('taskRunner fail');
-	tl.setResult(tl.TaskResult.Failed, tl.loc('GruntFailed', err.message));
-})
\ No newline at end of file
+	.then(function (code) {
+		publishTestResults(publishJUnitResults, testResultsFiles);
+		if (isCodeCoverageEnabled) {
+			npm.exec().then(function () {
+				istanbul.exec().then(function (code) {
+					publishCodeCoverage(summaryFile);
+					tl.setResult(tl.TaskResult.Succeeded, tl.loc('GruntReturnCode', code));
+				})
+					.fail(function (err) {
+						tl.debug('taskRunner fail');
+						tl.setResult(tl.TaskResult.Failed, tl.loc('IstanbulFailed', err.message));
+					});
+			}).fail(function (err) {
+				tl.debug('taskRunner fail');
+				tl.setResult(tl.TaskResult.Failed, tl.loc('NpmFailed', err.message));
+			})
+		} else {
+			tl.setResult(tl.TaskResult.Succeeded, tl.loc('GruntReturnCode', code));
+		}
+	})
+	.fail(function (err) {
+		publishTestResults(publishJUnitResults, testResultsFiles);
+		tl.debug('taskRunner fail');
+		tl.setResult(tl.TaskResult.Failed, tl.loc('GruntFailed', err.message));
+	})
+
+function publishTestResults(publishJUnitResults, testResultsFiles: string) {
```

Comment:
these function looks duplicate across grunt and gulp. can we have them in some common place like lib? check with nitin